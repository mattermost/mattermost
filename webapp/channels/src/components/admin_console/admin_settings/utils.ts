// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {AdminConfig, EnvironmentConfig} from '@mattermost/types/config';

function getConfigValue(config: AdminConfig | Partial<EnvironmentConfig>, path: string) {
    const pathParts = path.split('.');

    return pathParts.reduce((obj: object | null, pathPart) => {
        if (!obj) {
            return null;
        }
        return obj[(pathPart as keyof object)];
    }, config);
}

export function isSetByEnv(environmentConfig: Partial<EnvironmentConfig>, path: string) {
    return Boolean(getConfigValue(environmentConfig, path));
}

export const parseIntNonNegative = (str: string | number, defaultValue?: number) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < 0) {
        if (defaultValue) {
            return defaultValue;
        }
        return 0;
    }

    return n;
};

export const parseIntZeroOrMin = (str: string | number, minimumValue = 1) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < 0) {
        return 0;
    }
    if (n > 0 && n < minimumValue) {
        return minimumValue;
    }

    return n;
};

export const parseIntNonZero = (str: string | number, defaultValue?: number, minimumValue = 1) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < minimumValue) {
        if (defaultValue) {
            return defaultValue;
        }
        return 1;
    }

    return n;
};
