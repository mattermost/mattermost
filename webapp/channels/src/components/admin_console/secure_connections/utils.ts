// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {LocationDescriptor} from 'history';
import {DateTime, Interval} from 'luxon';
import {useCallback, useEffect, useMemo, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import type {ClientError} from '@mattermost/client';
import type {Channel} from '@mattermost/types/channels';
import {isRemoteClusterPatch, type RemoteCluster, type RemoteClusterPatch} from '@mattermost/types/remote_clusters';
import type {SharedChannelRemote} from '@mattermost/types/shared_channels';
import type {Team} from '@mattermost/types/teams';
import type {IDMappedObjects, RelationOneToOne} from '@mattermost/types/utilities';

import {ChannelTypes} from 'mattermost-redux/action_types';
import {getChannel as fetchChannel} from 'mattermost-redux/actions/channels';
import {Client4} from 'mattermost-redux/client';
import {getChannel} from 'mattermost-redux/selectors/entities/channels';
import {getActiveTeamsList, getTeam} from 'mattermost-redux/selectors/entities/teams';

import type {ActionFuncAsync} from 'types/store';

export const useRemoteClusters = () => {
    const [remoteClusters, setRemoteClusters] = useState<RemoteCluster[]>();
    const [loadingState, setLoadingState] = useState<boolean | ClientError>(true);
    const {loading, error} = loadingStatus(loadingState);

    const fetch = async () => {
        setLoadingState(true);
        try {
            const data = await Client4.getRemoteClusters({excludePlugins: true});
            setRemoteClusters(data?.length ? data : undefined);
            setLoadingState(false);
        } catch (err) {
            setLoadingState(err);
        }
    };

    useEffect(() => {
        fetch();
    }, []);

    return [remoteClusters, {loading, fetch, error}] as const;
};

export const useRemoteClusterEdit = (remoteId: string | 'create', initRemoteCluster: RemoteCluster | undefined) => {
    const editing = remoteId !== 'create';

    const [currentRemoteCluster, setCurrentRemoteCluster] = useState<RemoteCluster | undefined>(initRemoteCluster);
    const [patch, setPatch] = useState<Partial<RemoteClusterPatch>>({});
    const [loading, setLoading] = useState<TLoadingState>(editing && !currentRemoteCluster);
    const [saving, setSaving] = useState<TLoadingState>(false);

    const hasChanges = Object.keys(patch).length > 0;

    useEffect(() => {
        if (!editing) {
            return;
        }
        (async () => {
            try {
                const data = await Client4.getRemoteCluster(remoteId);
                setCurrentRemoteCluster(data);
                setLoading(false);
                setPatch({});
            } catch (err) {
                setSaving(err);
            }
        })();
    }, [remoteId]);

    const applyPatch = (patch: Partial<RemoteClusterPatch>) => {
        setPatch((current) => ({...current, ...patch}));
    };

    const save = async () => {
        if (currentRemoteCluster && isRemoteClusterPatch(patch)) {
            setSaving(true);
            try {
                const data = await Client4.patchRemoteCluster(remoteId, patch);
                setCurrentRemoteCluster(data);
                setSaving(false);
                setPatch({});
            } catch (err) {
                setSaving(err);
            }
            setSaving(false);
        }
    };

    return [
        {...currentRemoteCluster, ...patch},
        {
            applyPatch,
            save,
            hasChanges,
            loading,
            saving,
            currentRemoteCluster,
            patch,
        },
    ] as const;
};

export const useSharedChannelRemotes = (remoteId: string) => {
    const [remotes, setRemotes] = useState<RelationOneToOne<Channel, SharedChannelRemote>>();
    const [loadingState, setLoadingState] = useState<TLoadingState>(true);

    const loading = isPendingState(loadingState);
    const error = !loading && loadingState;

    const fetch = async () => {
        setLoadingState(true);
        try {
            const data = await Client4.getSharedChannelRemotes(remoteId, {include_deleted: true, include_unconfirmed: true});

            setRemotes(data?.reduce<typeof remotes>((state, remote) => {
                state![remote.channel_id] = remote;

                return state;
            }, {}));
            setLoadingState(false);
        } catch (error) {
            setLoadingState(error);
        }
    };

    useEffect(() => {
        fetch();
    }, [remoteId]);

    return [remotes, {loading, error, fetch}] as const;
};

const remoteRow = (remote: SharedChannelRemote, channel: Channel, team?: Team) => {
    return {...remote, display_name: channel.display_name, team_display_name: team?.display_name ?? ''};
};

export type SharedChannelRemoteRow = SharedChannelRemote & Pick<Channel, 'display_name'> & {team_display_name: string};
export const useSharedChannelRemoteRows = (remoteId: string, opts: {filter: 'home' | 'remote' | undefined}) => {
    const [sharedChannelRemotes, setSharedChannelRemotes] = useState<SharedChannelRemoteRow[]>();
    const [loadingState, setLoadingState] = useState<TLoadingState>(true);
    const dispatch = useDispatch();

    const loading = isPendingState(loadingState);
    const error = !loading && loadingState;

    const fetch = useCallback(async () => {
        if (opts.filter === undefined) {
            // wait for a filter
            return;
        }

        setLoadingState(true);
        dispatch<ActionFuncAsync<IDMappedObjects<SharedChannelRemoteRow>>>(async (dispatch, getState) => {
            const collected: IDMappedObjects<SharedChannelRemoteRow> = {};
            const missing: SharedChannelRemote[] = [];

            try {
                const data = await Client4.getSharedChannelRemotes(remoteId, {include_unconfirmed: true, exclude_remote: opts.filter === 'home', exclude_home: opts.filter === 'remote'});
                let state = getState();
                let getMyChannelsOnce: undefined | ((remote: SharedChannelRemote) => Promise<Channel | undefined>) = async (firstNotFound: SharedChannelRemote) => {
                    const channels = await Client4.getAllTeamsChannels();
                    dispatch({
                        type: ChannelTypes.RECEIVED_ALL_CHANNELS,
                        data: channels,
                    });
                    state = getState();

                    getMyChannelsOnce = undefined; // once-only

                    // return triggering not-found remote
                    return getChannel(state, firstNotFound.channel_id);
                };

                // first-past, get known channels or do initial load on first not-found channel
                for (const remote of data) {
                    // eslint-disable-next-line no-await-in-loop
                    const channel = getChannel(state, remote.channel_id) ?? await getMyChannelsOnce?.(remote);

                    if (!channel) {
                        // collect all remotes with missing channels
                        missing?.push(remote);
                        continue;
                    }

                    const team = getTeam(state, channel.team_id);
                    collected[remote.id] = remoteRow(remote, channel, team);
                }

                // fetch missing channels individually
                if (missing.length) {
                    // TODO: performance; consider adding sharedchannelremotes search param to api/v4/channels
                    await Promise.allSettled(missing.map((remote) => dispatch(fetchChannel(remote.channel_id))));
                    state = getState();

                    for (const remote of missing) {
                        const channel = getChannel(state, remote.channel_id);

                        if (!channel) {
                            continue;
                        }

                        const team = getTeam(state, channel.team_id);
                        collected[remote.id] = remoteRow(remote, channel, team);
                    }
                }

                const rows = Object.values(collected);

                setSharedChannelRemotes(rows.length ? rows : undefined);
                setLoadingState(false);
            } catch (error) {
                setLoadingState(error);
                return {error};
            }

            return {data: collected};
        });
    }, [remoteId, opts.filter]);

    useEffect(() => {
        fetch();
    }, [remoteId, opts.filter]);

    return [sharedChannelRemotes, {loading, error, fetch}] as const;
};

export const useTeamOptions = () => {
    const teams = useSelector(getActiveTeamsList);
    const teamsById = useMemo(() => teams.reduce<IDMappedObjects<Team>>((teams, team) => ({...teams, [team.id]: team}), {}), [teams]);
    return teamsById;
};

export const getEditLocation = (rc: RemoteCluster): LocationDescriptor<RemoteCluster> => {
    return {pathname: `/admin_console/site_config/secure_connections/${rc.remote_id}`, state: rc};
};

export const getCreateLocation = (): LocationDescriptor<RemoteCluster> => {
    return {pathname: '/admin_console/site_config/secure_connections/create'};
};

const SiteURLPendingPrefix = 'pending_';
export const isConfirmed = (rc: RemoteCluster) => Boolean(rc.site_url && !rc.site_url.startsWith(SiteURLPendingPrefix));
export const isConnected = (rc: RemoteCluster) => Interval.before(DateTime.now(), {minutes: 5}).contains(DateTime.fromMillis(rc.last_ping_at));

export type TLoadingState<TError extends Error = ClientError> = boolean | TError;
export const isPendingState = <T extends Error>(loadingState: TLoadingState<T>) => loadingState === true;
export const isErrorState = <T extends Error>(loadingState: TLoadingState<T>): loadingState is T => loadingState instanceof Error;

const loadingStatus = <T extends Error>(loadingState: TLoadingState<T>) => {
    const loading = isPendingState(loadingState);
    const error = isErrorState(loadingState) ? loadingState : undefined;

    return {error, loading};
};
