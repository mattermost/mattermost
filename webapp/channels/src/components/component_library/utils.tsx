// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

function buildPropsLists(inputPossibilities: {[x: string]: any[]}): Array<{[x: string]: any}> {
    const keys = Object.keys(inputPossibilities);
    if (!keys.length) {
        return [{}];
    }

    const selectedKey = keys[0];
    const restPossibilities = {...inputPossibilities};
    delete restPossibilities[selectedKey];
    const subProps = buildPropsLists(restPossibilities);
    const result: Array<{[x: string]: any}> = [];
    inputPossibilities[selectedKey].forEach((v) => {
        subProps.forEach((rest) => {
            result.push({...rest, [selectedKey]: v});
        });
    });

    return result;
}

function buildPropString(inputProps: {[x: string]: any}) {
    const propKeys = Object.keys(inputProps);
    if (!propKeys.length) {
        return undefined;
    }

    const result = [(<>{'PROPS: '}</>)];
    propKeys.forEach((v) => {
        result.push((<><b>{v}</b>{`: ${inputProps[v]}, `}</>));
    });
    return result;
}

export function buildComponent(
    Component: React.ComponentType<any>,
    propPossibilities: {[x: string]: any[]},
    dropdownPossibilities: Array<{[x: string]: string[]} | undefined>,
    setProps: Array<{[x: string]: any} | undefined>,
) {
    const res: React.ReactNode[] = [];
    let currentPropPossibilities = {...propPossibilities};
    dropdownPossibilities.forEach((v) => {
        if (v) {
            currentPropPossibilities = {
                ...currentPropPossibilities,
                ...v,
            };
        }
    });

    const propsVariations = buildPropsLists(currentPropPossibilities);
    let builtSetProps = {};
    setProps.forEach((v) => {
        if (v) {
            builtSetProps = {
                ...builtSetProps,
                ...v,
            };
        }
    });
    propsVariations.forEach((v) => {
        const propString = buildPropString(v);
        res.push(
            <>
                {Boolean(propString) && <p>{propString}</p>}
                <Component
                    {...builtSetProps}
                    {...v}
                />
            </>,
        );
    });
    return res;
}
