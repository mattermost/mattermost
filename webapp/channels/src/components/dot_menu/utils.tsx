// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type React from 'react';

export type ChangeEvent = React.KeyboardEvent | React.MouseEvent;

export const trackDotMenuEvent = (): void => {
    // Telemetry tracking removed
};
