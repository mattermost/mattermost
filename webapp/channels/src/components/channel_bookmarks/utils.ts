// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {useEffect, useMemo, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import type {Channel} from '@mattermost/types/channels';
import type {GlobalState} from '@mattermost/types/store';

import {Permissions} from 'mattermost-redux/constants';
import {getChannelBookmarks} from 'mattermost-redux/selectors/entities/channel_bookmarks';
import {getChannel, getMyChannelMember} from 'mattermost-redux/selectors/entities/channels';
import {getConfig, getFeatureFlagValue, getLicense} from 'mattermost-redux/selectors/entities/general';
import {haveIChannelPermission} from 'mattermost-redux/selectors/entities/roles';
import {insertWithoutDuplicates} from 'mattermost-redux/utils/array_utils';

import {fetchChannelBookmarks, reorderBookmark} from 'actions/channel_bookmarks';
import {loadCustomEmojisIfNeeded} from 'actions/emoji_actions';

import Constants from 'utils/constants';
import {trimmedEmojiName} from 'utils/emoji_utils';
import {canUploadFiles, isPublicLinksEnabled} from 'utils/file_utils';

export const MAX_BOOKMARKS_PER_CHANNEL = 50;

const {OPEN_CHANNEL, PRIVATE_CHANNEL, GM_CHANNEL, DM_CHANNEL} = Constants as {OPEN_CHANNEL: 'O'; PRIVATE_CHANNEL: 'P'; GM_CHANNEL: 'G'; DM_CHANNEL: 'D'};

type TAction = 'add' | 'edit' | 'delete' | 'order';
type TActionKey = `${TAction}${typeof OPEN_CHANNEL | typeof PRIVATE_CHANNEL}`;

const key = (a: TAction, c: typeof OPEN_CHANNEL | typeof PRIVATE_CHANNEL): TActionKey => {
    return `${a}${c}`;
};

const BOOKMARK_PERMISSION = {

    // open channel
    [key('add', OPEN_CHANNEL)]: Permissions.ADD_BOOKMARK_PUBLIC_CHANNEL,
    [key('edit', OPEN_CHANNEL)]: Permissions.EDIT_BOOKMARK_PUBLIC_CHANNEL,
    [key('delete', OPEN_CHANNEL)]: Permissions.DELETE_BOOKMARK_PUBLIC_CHANNEL,
    [key('order', OPEN_CHANNEL)]: Permissions.ORDER_BOOKMARK_PUBLIC_CHANNEL,

    // private channel
    [key('add', PRIVATE_CHANNEL)]: Permissions.ADD_BOOKMARK_PRIVATE_CHANNEL,
    [key('edit', PRIVATE_CHANNEL)]: Permissions.EDIT_BOOKMARK_PRIVATE_CHANNEL,
    [key('delete', PRIVATE_CHANNEL)]: Permissions.DELETE_BOOKMARK_PRIVATE_CHANNEL,
    [key('order', PRIVATE_CHANNEL)]: Permissions.ORDER_BOOKMARK_PRIVATE_CHANNEL,
} as const;

export const useChannelBookmarkPermission = (channelId: string, action: TAction) => {
    return useSelector((state: GlobalState) => getHaveIChannelBookmarkPermission(state, channelId, action));
};

export const getHaveIChannelBookmarkPermission = (state: GlobalState, channelId: string, action: TAction) => {
    const channel: Channel | undefined = getChannel(state, channelId);

    if (!channel) {
        return false;
    }

    if (channel.delete_at !== 0) {
        return false;
    }

    const {type} = channel;

    if (type === 'threads') {
        return false;
    }

    if (type === GM_CHANNEL || type === DM_CHANNEL) {
        const myMembership = getMyChannelMember(state, channelId);
        return myMembership?.channel_id === channelId;
    }

    const permission = BOOKMARK_PERMISSION[key(action, type)];

    return channel && permission && haveIChannelPermission(state, channel.team_id, channelId, permission);
};

export const useCanUploadFiles = () => {
    return useSelector((state: GlobalState) => canUploadFiles(getConfig(state)));
};

export const useCanGetPublicLink = () => {
    return useSelector((state: GlobalState) => isPublicLinksEnabled(getConfig(state)));
};

export const useCanGetLinkPreviews = () => {
    return useSelector((state: GlobalState) => getConfig(state).EnableLinkPreviews === 'true');
};

export const getIsChannelBookmarksEnabled = (state: GlobalState) => {
    const isEnabled = getFeatureFlagValue(state, 'ChannelBookmarks') === 'true';

    if (!isEnabled) {
        return false;
    }

    const license = getLicense(state);

    return license?.IsLicensed === 'true';
};

export const useChannelBookmarks = (channelId: string) => {
    const dispatch = useDispatch();
    const bookmarks = useSelector((state: GlobalState) => getChannelBookmarks(state, channelId));

    const order = useMemo(() => {
        return Object.keys(bookmarks).sort((a, b) => bookmarks[a].sort_order - bookmarks[b].sort_order);
    }, [bookmarks]);
    const [tempOrder, setTempOrder] = useState<typeof order>();

    useEffect(() => {
        if (tempOrder) {
            setTempOrder(undefined);
        }
    }, [order]);

    useEffect(() => {
        if (channelId) {
            dispatch(fetchChannelBookmarks(channelId));
        }
    }, [channelId]);

    useEffect(() => {
        const emojis = Object.values(bookmarks).reduce<string[]>((result, {emoji}) => {
            if (emoji) {
                result.push(trimmedEmojiName(emoji));
            }

            return result;
        }, []);

        if (emojis.length) {
            dispatch(loadCustomEmojisIfNeeded(emojis));
        }
    }, [bookmarks]);

    const reorder = async (id: string, prevOrder: number, nextOrder: number) => {
        setTempOrder(insertWithoutDuplicates(order, id, nextOrder));
        const {error} = await dispatch(reorderBookmark(channelId, id, nextOrder));

        if (error) {
            setTempOrder(undefined);
        }
    };

    return {
        bookmarks,
        order: tempOrder ?? order,
        reorder,
    } as const;
};

