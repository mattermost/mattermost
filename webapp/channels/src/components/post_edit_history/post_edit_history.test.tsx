// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import type {ComponentProps} from 'react';

import {Client4} from 'mattermost-redux/client';

import {renderWithContext, waitFor} from 'tests/react_testing_utils';
import {TestHelper} from 'utils/test_helper';

import PostEditHistory from './post_edit_history';

describe('components/post_edit_history', () => {
    const baseProps: ComponentProps<typeof PostEditHistory> = {
        channelDisplayName: 'channel_display_name',
        originalPost: TestHelper.getPostMock({
            id: 'post_id',
            message: 'post message',
        }),
        dispatch: jest.fn(),
    };
    const mock = jest.spyOn(Client4, 'getPostEditHistory');

    test('should match snapshot', async () => {
        const data = [
            TestHelper.getPostMock({
                id: 'post_id_1',
                message: 'post message version 1',
            }),
            TestHelper.getPostMock({
                id: 'post_id_2',
                message: 'post message version 2',
            }),
        ];
        mock.mockResolvedValue(data);

        const wrapper = await waitFor(() => {
            return renderWithContext(<PostEditHistory {...baseProps}/>);
        });

        expect(wrapper.container).toMatchSnapshot();
        expect(mock).toBeCalledWith(baseProps.originalPost.id);
    });

    test('should display error screen if errors are present', async () => {
        const error = new Error('An example error');
        mock.mockRejectedValue(error);

        const wrapper = await waitFor(() => {
            return renderWithContext(<PostEditHistory {...baseProps}/>);
        });

        expect(wrapper.container).toMatchSnapshot();
        expect(mock).toBeCalledWith(baseProps.originalPost.id);
    });
});
