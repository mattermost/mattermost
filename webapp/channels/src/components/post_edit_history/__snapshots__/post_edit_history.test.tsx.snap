// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/post_edit_history should display error screen if errors are present 1`] = `
<div>
  <div
    class="sidebar-right__body sidebar-right__edit-post-history"
    id="rhsContainer"
  >
    <div
      style="position: relative; overflow: hidden; width: 100%; height: 100%;"
    >
      <div
        class="scrollbar--view"
        style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; overflow: scroll; margin-right: 0px; margin-bottom: 0px;"
      >
        <div
          class="sidebar--right__header"
        >
          <span
            class="sidebar--right__title"
          >
            Edit History
            <div
              class="sidebar--right__title__channel"
            >
              channel_display_name
            </div>
          </span>
          <div
            class="pull-right"
          >
            <button
              aria-label="Expand Sidebar Icon"
              class="sidebar--right__expand btn btn-icon btn-sm"
              type="button"
            >
              <i
                aria-hidden="true"
                class="icon icon-arrow-expand"
              />
              <i
                aria-hidden="true"
                class="icon icon-arrow-collapse"
              />
            </button>
            <button
              aria-label="Close"
              class="sidebar--right__close btn btn-icon btn-sm"
              id="searchResultsCloseButton"
              type="button"
            >
              <i
                aria-label="Close Sidebar Icon"
                class="icon icon-close"
              />
            </button>
          </div>
        </div>
        <div
          class="edit-post-history__error_container"
        >
          <div
            class="edit-post-history__error_item"
          >
            <svg
              fill="none"
              height="127"
              viewBox="0 0 87 70"
              width="127"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="var(--button-bg)"
                fill-opacity="0.12"
                height="24"
                rx="3.75"
                width="72"
                x="8.00098"
                y="7"
              />
              <rect
                fill="var(--button-bg)"
                fill-opacity="0.12"
                height="25"
                rx="3.75"
                width="87"
                x="0.000976562"
                y="34"
              />
              <path
                d="M38.3214 2.31098C39.4303 0.112261 42.5697 0.112256 43.6786 2.31098L71.7146 57.899C72.7209 59.8943 71.2707 62.25 69.0359 62.25H12.9641C10.7294 62.25 9.27912 59.8943 10.2854 57.899L38.3214 2.31098Z"
                fill="var(--center-channel-bg)"
              />
              <path
                d="M40.3214 4.31098C41.4303 2.11226 44.5697 2.11226 45.6786 4.31098L73.7146 59.899C74.7209 61.8943 73.2707 64.25 71.0359 64.25H14.9641C12.7294 64.25 11.2791 61.8943 12.2854 59.899L40.3214 4.31098Z"
                fill="#FFBC1F"
              />
              <path
                d="M43.2322 2.53614L71.2681 58.1242C72.1067 59.7869 70.8982 61.75 69.0359 61.75H12.9641C11.1018 61.75 9.89327 59.7869 10.7319 58.1242L38.7678 2.53614C39.6919 0.703873 42.3081 0.703871 43.2322 2.53614Z"
                stroke="var(--center-channel-color)"
              />
              <path
                d="M49.542 4.23999L52.8888 10.72M74.922 53.38L68.5073 40.96L66.8339 37.72L64.6027 33.4L61.5348 27.46M59.3036 23.14L55.12 15.04"
                stroke="var(--center-channel-color)"
                stroke-linecap="round"
                stroke-opacity="0.56"
                stroke-width="1.08"
              />
              <path
                d="M38.0164 25.2833L40.2971 39.9301C40.3191 40.2208 40.4554 40.4927 40.6786 40.6912C40.9018 40.8897 41.1954 41 41.5002 41C41.8051 41 42.0986 40.8897 42.3219 40.6912C42.5451 40.4927 42.6814 40.2208 42.7034 39.9301L44.984 25.2833C45.3987 19.5722 37.5955 19.5722 38.0164 25.2833Z"
                fill="#3F4350"
              />
              <path
                d="M41.0072 47C41.798 47.0014 42.5706 47.2372 43.2275 47.6776C43.8843 48.118 44.396 48.7432 44.6976 49.4742C44.9993 50.2053 45.0774 51.0093 44.9222 51.7848C44.7671 52.5602 44.3856 53.2723 43.8259 53.831C43.2662 54.3897 42.5535 54.7699 41.7777 54.9237C41.002 55.0774 40.1981 54.9978 39.4676 54.6948C38.7371 54.3919 38.1128 53.8792 37.6736 53.2215C37.2344 52.5639 37 51.7908 37 51C37 50.4741 37.1036 49.9534 37.3051 49.4676C37.5066 48.9818 37.8019 48.5406 38.1741 48.169C38.5463 47.7975 38.9881 47.503 39.4743 47.3024C39.9604 47.1018 40.4813 46.9991 41.0072 47Z"
                fill="#3F4350"
              />
              <path
                d="M48.4619 68.5H70.0619"
                stroke="var(--center-channel-color)"
                stroke-linecap="round"
                stroke-opacity="0.56"
                stroke-width="1.08"
              />
              <path
                d="M10.001 50L26.001 19"
                stroke="var(--center-channel-color)"
                stroke-linecap="round"
                stroke-opacity="0.56"
                stroke-width="1.08"
              />
            </svg>
            <p
              class="edit-post-history__error_heading"
            >
              Unable to load edit history
            </p>
            <p
              class="edit-post-history__error_subheading"
            >
              There was an error loading the history for this message. Check your network connection or try again later.
            </p>
          </div>
        </div>
      </div>
      <div
        style="position: absolute; height: 6px; transition: opacity 500ms; opacity: 0; display: none; right: 2px; bottom: 2px; left: 2px; border-radius: 3px;"
      >
        <div
          class="scrollbar--horizontal"
          style="position: relative; display: block; height: 100%;"
        />
      </div>
      <div
        style="position: absolute; width: 6px; transition: opacity 500ms; opacity: 0; display: none; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
      >
        <div
          class="scrollbar--vertical"
          style="position: relative; display: block; width: 100%;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`components/post_edit_history should match snapshot 1`] = `
<div>
  <div
    class="sidebar-right__body sidebar-right__edit-post-history"
    id="rhsContainer"
  >
    <div
      style="position: relative; overflow: hidden; width: 100%; height: 100%;"
    >
      <div
        class="scrollbar--view"
        style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; overflow: scroll; margin-right: 0px; margin-bottom: 0px;"
      >
        <div
          class="sidebar--right__header"
        >
          <span
            class="sidebar--right__title"
          >
            Edit History
            <div
              class="sidebar--right__title__channel"
            >
              channel_display_name
            </div>
          </span>
          <div
            class="pull-right"
          >
            <button
              aria-label="Expand Sidebar Icon"
              class="sidebar--right__expand btn btn-icon btn-sm"
              type="button"
            >
              <i
                aria-hidden="true"
                class="icon icon-arrow-expand"
              />
              <i
                aria-hidden="true"
                class="icon icon-arrow-collapse"
              />
            </button>
            <button
              aria-label="Close"
              class="sidebar--right__close btn btn-icon btn-sm"
              id="searchResultsCloseButton"
              type="button"
            >
              <i
                aria-label="Close Sidebar Icon"
                class="icon icon-close"
              />
            </button>
          </div>
        </div>
        <div
          class="edit-post-history__container edit-post-history__container__background"
        >
          <div
            aria-label="At 12:00 AM Thursday, January 1, Someone wrote, post message"
            class="a11y__section post"
            id="searchResult_post_id"
          >
            <div
              class="edit-post-history__title__container"
            >
              <div
                class="edit-post-history__date__badge__container"
              >
                <button
                  aria-label="Toggle to see an old message."
                  class="edit-post-history__icon__button toggleCollapseButton"
                >
                  <i
                    class="icon icon-chevron-down"
                  />
                </button>
                <span
                  class="edit-post-history__date"
                >
                  <time
                    datetime="1970-01-01T00:00:00.000"
                  >
                    January 01, 1970 at 12:00 AM
                  </time>
                </span>
                <div
                  class="edit-post-history__current__indicator"
                >
                  Current Version
                </div>
              </div>
            </div>
            <div
              class="edit-post-history__content_container"
            >
              <div
                class="edit-post-history__header"
              >
                <span
                  class="profile-icon"
                >
                  <img
                    alt="user profile image"
                    class="Avatar Avatar-sm avatar-post-preview"
                    loading="lazy"
                    src="/api/v4/users/user_id/image?_=0"
                  />
                </span>
                <div
                  class="edit-post-history__header__username"
                >
                  <div
                    class="user-popover"
                  >
                    Someone
                  </div>
                </div>
              </div>
              <div
                class="post__content"
              >
                <div
                  class="search-item-snippet post__body"
                >
                  <div
                    class="post-message post-message--collapsed"
                  >
                    <div
                      class="post-message__text-container"
                      style="max-height: 600px;"
                    >
                      <div
                        class="post-message__text"
                        dir="auto"
                        id="rhsPostMessageText_post_id"
                        tabindex="0"
                      >
                        <p>
                          post message
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="edit-post-history__container"
        >
          <div
            aria-label="At 12:00 AM Thursday, January 1, Someone wrote, post message version 1"
            class="a11y__section post"
            id="searchResult_post_id_1"
          >
            <div
              class="edit-post-history__title__container"
            >
              <div
                class="edit-post-history__date__badge__container"
              >
                <button
                  aria-label="Toggle to see an old message."
                  class="edit-post-history__icon__button toggleCollapseButton"
                >
                  <i
                    class="icon icon-chevron-right"
                  />
                </button>
                <span
                  class="edit-post-history__date"
                >
                  <time
                    datetime="1970-01-01T00:00:00.000"
                  >
                    January 01, 1970 at 12:00 AM
                  </time>
                </span>
              </div>
              <button
                aria-label="Select to restore an old message."
                class="edit-post-history__icon__button restore-icon"
              >
                <i
                  class="icon icon-restore"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="edit-post-history__container"
        >
          <div
            aria-label="At 12:00 AM Thursday, January 1, Someone wrote, post message version 2"
            class="a11y__section post"
            id="searchResult_post_id_2"
          >
            <div
              class="edit-post-history__title__container"
            >
              <div
                class="edit-post-history__date__badge__container"
              >
                <button
                  aria-label="Toggle to see an old message."
                  class="edit-post-history__icon__button toggleCollapseButton"
                >
                  <i
                    class="icon icon-chevron-right"
                  />
                </button>
                <span
                  class="edit-post-history__date"
                >
                  <time
                    datetime="1970-01-01T00:00:00.000"
                  >
                    January 01, 1970 at 12:00 AM
                  </time>
                </span>
              </div>
              <button
                aria-label="Select to restore an old message."
                class="edit-post-history__icon__button restore-icon"
              >
                <i
                  class="icon icon-restore"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        style="position: absolute; height: 6px; transition: opacity 500ms; opacity: 0; display: none; right: 2px; bottom: 2px; left: 2px; border-radius: 3px;"
      >
        <div
          class="scrollbar--horizontal"
          style="position: relative; display: block; height: 100%;"
        />
      </div>
      <div
        style="position: absolute; width: 6px; transition: opacity 500ms; opacity: 0; display: none; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
      >
        <div
          class="scrollbar--vertical"
          style="position: relative; display: block; width: 100%;"
        />
      </div>
    </div>
  </div>
</div>
`;
